File: native-android/app/src/main/kotlin/com/crossdashboard/app/data/parser/ICalParser.kt
package com.crossdashboard.app.data.parser

import com.crossdashboard.app.domain.model.*
import java.time.*
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * Hand-written iCalendar (RFC 5545) parser for VEVENT, VTODO, and VJOURNAL components.
 *
 * Handles:
 * - Line unfolding (CRLF + whitespace continuation)
 * - Quoted-printable / base64 value decoding
 * - DTSTART / DTEND / DUE with TZID param and UTC Z suffix
 * - All-day dates (VALUE=DATE, no time component)
 * - RELATED-TO;RELTYPE=PARENT for subtask hierarchy
 * - CATEGORIES comma-separated list
 */
object ICalParser {

    // ─── Public entry points ─────────────────────────────────────────────────

    fun parseEvents(
        icalText: String,
        calendarHref: String? = null,
        resourceHref: String? = null,
        resourceEtag: String? = null,
    ): List<CalendarEvent> {
        val events = mutableListOf<CalendarEvent>()
        forEachComponent(icalText, "VEVENT") { props ->
            parseEvent(props, calendarHref, resourceHref, resourceEtag)?.let { events.add(it) }
        }
        return events
    }

    fun parseTasks(
        icalText: String,
        calendarHref: String? = null,
        resourceHref: String? = null,
        resourceEtag: String? = null,
    ): List<CalDavTask> {
        val tasks = mutableListOf<CalDavTask>()
        forEachComponent(icalText, "VTODO") { props ->
            parseTask(props, calendarHref, resourceHref, resourceEtag)?.let { tasks.add(it) }
        }
        return tasks
    }

    fun parseNotes(
        icalText: String,
        calendarHref: String? = null,
        resourceHref: String? = null,
        resourceEtag: String? = null,
    ): List<Note> {
        val notes = mutableListOf<Note>()
        forEachComponent(icalText, "VJOURNAL") { props ->
            parseNote(props, calendarHref, resourceHref, resourceEtag)?.let { notes.add(it) }
        }
        return notes
    }

    // ─── Component iterator ──────────────────────────────────────────────────

    private fun forEachComponent(
        icalText: String,
        componentName: String,
        block: (Map<String, String>) -> Unit,
    ) {
        val lines = unfold(icalText)
        var inside = false
        val props = mutableMapOf<String, String>()

        for (line in lines) {
            when {
                line.equals("BEGIN:$componentName", ignoreCase = true) -> {
                    inside = true
                    props.clear()
                }
                line.equals("END:$componentName", ignoreCase = true) && inside -> {
                    inside = false
                    block(props.toMap())
                }
                inside -> {
                    val colonIdx = line.indexOf(':')
                    if (colonIdx > 0) {
                        // Extract property name (strip params like ;TZID=...)
                        val rawName = line.substring(0, colonIdx)
                        val value = line.substring(colonIdx + 1)
                        val normalizedName = normalizePropertyName(rawName)
                        // Preserve first occurrence for most props; append for multi-value
                        if (normalizedName == "RELATED-TO" || normalizedName == "ATTENDEE") {
                            val existing = props[normalizedName]
                            props[normalizedName] = if (existing == null) value else "$existing,$value"
                        } else {
                            props.putIfAbsent(normalizedName, value)
                        }
                        // Also store full param line for DTSTART/DTEND/DUE so we can get TZID
                        props["__RAW_$normalizedName"] = line
                    }
                }
            }
        }
    }

    private fun normalizePropertyName(rawName: String): String {
        // "DTSTART;TZID=America/New_York" → "DTSTART"
        val semiIdx = rawName.indexOf(';')
        return if (semiIdx > 0) rawName.substring(0, semiIdx).uppercase()
        else rawName.uppercase()
    }

    // ─── Event parser ─────────────────────────────────────────────────────────

    private fun parseEvent(
        props: Map<String, String>,
        calendarHref: String?,
        resourceHref: String? = null,
        resourceEtag: String? = null,
    ): CalendarEvent? {
        val uid = props["UID"] ?: return null
        val summary = unescape(props["SUMMARY"] ?: return null)
        val start = parseDateTime(props, "DTSTART") ?: return null
        val end = parseDateTime(props, "DTEND")
            ?: parseDateTime(props, "DURATION")?.let { start.plusMillis(parseDuration(props["DURATION"])) }
            ?: start.plus(Duration.ofHours(1))
        return CalendarEvent(
            uid = uid,
            summary = summary,
            start = start,
            end = end,
            description = props["DESCRIPTION"]?.let { unescape(it) },
            location = props["LOCATION"]?.let { unescape(it) },
            calendarHref = calendarHref,
            etag = resourceEtag,
            href = resourceHref,
        )
    }

    // ─── Task parser ──────────────────────────────────────────────────────────

    private fun parseTask(
        props: Map<String, String>,
        calendarHref: String?,
        resourceHref: String? = null,
        resourceEtag: String? = null,
    ): CalDavTask? {
        val uid = props["UID"] ?: return null
        val summary = unescape(props["SUMMARY"] ?: return null)

        val relatedTo = props["RELATED-TO"]
        val parentUid = if (relatedTo != null) {
            // May be a comma-separated list; pick the PARENT type
            val rawLine = props["__RAW_RELATED-TO"] ?: ""
            if (rawLine.contains("RELTYPE=PARENT", ignoreCase = true)) {
                relatedTo.split(",").firstOrNull()
            } else null
        } else null

        return CalDavTask(
            uid = uid,
            summary = summary,
            description = props["DESCRIPTION"]?.let { unescape(it) },
            status = TaskStatus.fromIcal(props["STATUS"] ?: "NEEDS-ACTION"),
            priority = props["PRIORITY"]?.toIntOrNull() ?: 0,
            percentComplete = props["PERCENT-COMPLETE"]?.toIntOrNull() ?: 0,
            due = parseDateTime(props, "DUE"),
            dtstart = parseDateTime(props, "DTSTART"),
            completed = parseDateTime(props, "COMPLETED"),
            created = parseDateTime(props, "CREATED") ?: Instant.now(),
            lastModified = parseDateTime(props, "LAST-MODIFIED") ?: Instant.now(),
            categories = parseCategories(props["CATEGORIES"]),
            location = props["LOCATION"]?.let { unescape(it) },
            parentUid = parentUid,
            calendarHref = calendarHref,
            etag = resourceEtag,
            href = resourceHref,
        )
    }

    // ─── Note (VJOURNAL) parser ───────────────────────────────────────────────

    private fun parseNote(
        props: Map<String, String>,
        calendarHref: String?,
        resourceHref: String? = null,
        resourceEtag: String? = null,
    ): Note? {
        val uid = props["UID"] ?: return null
        val summary = unescape(props["SUMMARY"] ?: return null)
        return Note(
            uid = uid,
            summary = summary,
            body = props["DESCRIPTION"]?.let { unescape(it) } ?: "",
            categories = parseCategories(props["CATEGORIES"]),
            created = parseDateTime(props, "DTSTAMP") ?: Instant.now(),
            lastModified = parseDateTime(props, "LAST-MODIFIED") ?: Instant.now(),
            calendarHref = calendarHref,
            etag = resourceEtag,
            href = resourceHref,
        )
    }

    // ─── iCal serializers (domain → iCal text) ───────────────────────────────

    fun serializeTask(task: CalDavTask): String = buildString {
        appendLine("BEGIN:VCALENDAR")
        appendLine("VERSION:2.0")
        appendLine("PRODID:-//CrossDashboard//Native//EN")
        appendLine("BEGIN:VTODO")
        appendLine("UID:${task.uid}")
        appendLine("DTSTAMP:${formatInstant(Instant.now())}")
        appendLine("SUMMARY:${escape(task.summary)}")
        task.description?.let { appendLine("DESCRIPTION:${escape(it)}") }
        appendLine("STATUS:${task.status.icalValue}")
        appendLine("PRIORITY:${task.priority}")
        appendLine("PERCENT-COMPLETE:${task.percentComplete}")
        task.due?.let { appendLine("DUE:${formatInstant(it)}") }
        task.dtstart?.let { appendLine("DTSTART:${formatInstant(it)}") }
        task.completed?.let { appendLine("COMPLETED:${formatInstant(it)}") }
        appendLine("CREATED:${formatInstant(task.created)}")
        appendLine("LAST-MODIFIED:${formatInstant(Instant.now())}")
        if (task.categories.isNotEmpty()) {
            appendLine("CATEGORIES:${task.categories.joinToString(",")}")
        }
        task.location?.let { appendLine("LOCATION:${escape(it)}") }
        task.parentUid?.let { appendLine("RELATED-TO;RELTYPE=PARENT:$it") }
        appendLine("END:VTODO")
        appendLine("END:VCALENDAR")
    }

    fun serializeNote(note: Note): String = buildString {
        appendLine("BEGIN:VCALENDAR")
        appendLine("VERSION:2.0")
        appendLine("PRODID:-//CrossDashboard//Native//EN")
        appendLine("BEGIN:VJOURNAL")
        appendLine("UID:${note.uid}")
        appendLine("SUMMARY:${escape(note.summary)}")
        if (note.body.isNotBlank()) appendLine("DESCRIPTION:${escape(note.body)}")
        if (note.categories.isNotEmpty()) appendLine("CATEGORIES:${note.categories.joinToString(",")}")
        appendLine("DTSTAMP:${formatInstant(Instant.now())}")
        appendLine("LAST-MODIFIED:${formatInstant(Instant.now())}")
        appendLine("END:VJOURNAL")
        appendLine("END:VCALENDAR")
    }

    fun serializeEvent(event: CalendarEvent): String = buildString {
        appendLine("BEGIN:VCALENDAR")
        appendLine("VERSION:2.0")
        appendLine("PRODID:-//CrossDashboard//Native//EN")
        appendLine("BEGIN:VEVENT")
        appendLine("UID:${event.uid}")
        appendLine("SUMMARY:${escape(event.summary)}")
        appendLine("DTSTART:${formatInstant(event.start)}")
        appendLine("DTEND:${formatInstant(event.end)}")
        event.description?.let { appendLine("DESCRIPTION:${escape(it)}") }
        event.location?.let { appendLine("LOCATION:${escape(it)}") }
        appendLine("DTSTAMP:${formatInstant(Instant.now())}")
        appendLine("LAST-MODIFIED:${formatInstant(Instant.now())}")
        appendLine("END:VEVENT")
        appendLine("END:VCALENDAR")
    }

    // ─── Parsing helpers ─────────────────────────────────────────────────────

    private fun unfold(icalText: String): List<String> {
        val lines = mutableListOf<String>()
        val sb = StringBuilder()
        for (line in icalText.lines()) {
            when {
                line.startsWith(' ') || line.startsWith('\t') -> sb.append(line.substring(1))
                else -> {
                    if (sb.isNotEmpty()) lines.add(sb.toString())
                    sb.clear()
                    sb.append(line)
                }
            }
        }
        if (sb.isNotEmpty()) lines.add(sb.toString())
        return lines
    }

    private fun parseDateTime(props: Map<String, String>, key: String): Instant? {
        val rawLine = props["__RAW_$key"] ?: return null
        val colonIdx = rawLine.indexOf(':')
        if (colonIdx < 0) return null
        val value = rawLine.substring(colonIdx + 1).trim()
        val paramPart = rawLine.substring(0, colonIdx)

        // VALUE=DATE → all-day, set to midnight UTC
        if (paramPart.contains("VALUE=DATE", ignoreCase = true)) {
            return try {
                LocalDate.parse(value, DateTimeFormatter.BASIC_ISO_DATE)
                    .atStartOfDay(ZoneOffset.UTC).toInstant()
            } catch (_: DateTimeParseException) { null }
        }

        // UTC Z suffix
        if (value.endsWith("Z")) {
            return try {
                Instant.parse(value.take(15).let {
                    "${it.substring(0, 4)}-${it.substring(4, 6)}-${it.substring(6, 8)}T" +
                        "${it.substring(9, 11)}:${it.substring(11, 13)}:${it.substring(13, 15)}Z"
                })
            } catch (_: Exception) {
                runCatching {
                    LocalDateTime.parse(value.dropLast(1), DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss"))
                        .toInstant(ZoneOffset.UTC)
                }.getOrNull()
            }
        }

        // TZID param
        val tzidMatch = Regex("TZID=([^;:]+)").find(paramPart)
        val tz = tzidMatch?.groupValues?.get(1)?.let {
            runCatching { ZoneId.of(it) }.getOrNull() ?: ZoneOffset.UTC
        } ?: ZoneOffset.UTC

        return try {
            LocalDateTime.parse(value, DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss"))
                .atZone(tz).toInstant()
        } catch (_: DateTimeParseException) { null }
    }

    private fun parseDuration(durationStr: String?): Long {
        if (durationStr == null) return 3600_000L
        var ms = 0L
        val m = Regex("P(?:(\\d+)W)?(?:(\\d+)D)?(?:T(?:(\\d+)H)?(?:(\\d+)M)?(?:(\\d+)S)?)?")
            .find(durationStr) ?: return 3600_000L
        ms += (m.groupValues[1].toLongOrNull() ?: 0) * 7 * 24 * 3600_000
        ms += (m.groupValues[2].toLongOrNull() ?: 0) * 24 * 3600_000
        ms += (m.groupValues[3].toLongOrNull() ?: 0) * 3600_000
        ms += (m.groupValues[4].toLongOrNull() ?: 0) * 60_000
        ms += (m.groupValues[5].toLongOrNull() ?: 0) * 1000
        return ms.takeIf { it > 0 } ?: 3600_000L
    }

    private fun parseCategories(raw: String?): List<String> {
        if (raw.isNullOrBlank()) return emptyList()
        return raw.split(",").map { it.trim() }.filter { it.isNotEmpty() }
    }

    private fun unescape(value: String): String =
        value
            .replace("\\n", "\n")
            .replace("\\N", "\n")
            .replace("\\,", ",")
            .replace("\\;", ";")
            .replace("\\\\", "\\")

    private fun escape(value: String): String =
        value
            .replace("\\", "\\\\")
            .replace("\n", "\\n")
            .replace(",", "\\,")
            .replace(";", "\\;")

    private val UTC_FORMATTER = DateTimeFormatter
        .ofPattern("yyyyMMdd'T'HHmmss'Z'")
        .withZone(ZoneOffset.UTC)

    private fun formatInstant(instant: Instant): String =
        UTC_FORMATTER.format(instant)
}

File: native-macos/CrossDashboard/Data/Parser/ICalParser.swift
import Foundation
import CrossDashboardKit

/// Hand-written iCalendar (RFC 5545) parser for VEVENT, VTODO, and VJOURNAL components.
///
/// Handles:
/// - Line unfolding (CRLF + whitespace continuation)
/// - DTSTART / DTEND / DUE with TZID param and UTC Z suffix
/// - All-day dates (VALUE=DATE, no time component)
/// - RELATED-TO;RELTYPE=PARENT for subtask hierarchy
/// - CATEGORIES comma-separated list
///
/// Direct Swift port of ICalParser.kt.
enum ICalParser {

    // ─── Public entry points ─────────────────────────────────────────────────

    static func parseEvents(
        icalText: String,
        calendarHref: String? = nil,
        resourceHref: String? = nil,
        resourceEtag: String? = nil
    ) -> [CalendarEvent] {
        var results: [CalendarEvent] = []
        forEachComponent(in: icalText, named: "VEVENT") { props in
            if let event = parseEvent(props, calendarHref: calendarHref, resourceHref: resourceHref, resourceEtag: resourceEtag) {
                results.append(event)
            }
        }
        return results
    }

    static func parseTasks(
        icalText: String,
        calendarHref: String? = nil,
        resourceHref: String? = nil,
        resourceEtag: String? = nil
    ) -> [CalDavTask] {
        var results: [CalDavTask] = []
        forEachComponent(in: icalText, named: "VTODO") { props in
            if let task = parseTask(props, calendarHref: calendarHref, resourceHref: resourceHref, resourceEtag: resourceEtag) {
                results.append(task)
            }
        }
        return results
    }

    static func parseNotes(
        icalText: String,
        calendarHref: String? = nil,
        resourceHref: String? = nil,
        resourceEtag: String? = nil
    ) -> [Note] {
        var results: [Note] = []
        forEachComponent(in: icalText, named: "VJOURNAL") { props in
            if let note = parseNote(props, calendarHref: calendarHref, resourceHref: resourceHref, resourceEtag: resourceEtag) {
                results.append(note)
            }
        }
        return results
    }

    // ─── Component iterator ──────────────────────────────────────────────────

    private static func forEachComponent(
        in icalText: String,
        named componentName: String,
        block: ([String: String]) -> Void
    ) {
        let lines = unfold(icalText)
        var inside = false
        var props: [String: String] = [:]
        let beginMarker = "BEGIN:\(componentName)".uppercased()
        let endMarker   = "END:\(componentName)".uppercased()

        for line in lines {
            let upper = line.uppercased()
            if upper == beginMarker {
                inside = true
                props = [:]
            } else if upper == endMarker && inside {
                inside = false
                block(props)
            } else if inside {
                if let colonIdx = line.firstIndex(of: ":") {
                    let rawName = String(line[line.startIndex ..< colonIdx])
                    let value   = String(line[line.index(after: colonIdx)...])
                    let name    = normalizePropertyName(rawName)

                    if name == "RELATED-TO" || name == "ATTENDEE" {
                        let existing = props[name]
                        props[name] = existing == nil ? value : "\(existing!),\(value)"
                    } else if props[name] == nil {
                        props[name] = value
                    }
                    // Store raw line for TZID extraction
                    props["__RAW_\(name)"] = line
                }
            }
        }
    }

    private static func normalizePropertyName(_ rawName: String) -> String {
        if let semiIdx = rawName.firstIndex(of: ";") {
            return String(rawName[rawName.startIndex ..< semiIdx]).uppercased()
        }
        return rawName.uppercased()
    }

    // ─── Event parser ─────────────────────────────────────────────────────────

    private static func parseEvent(
        _ props: [String: String],
        calendarHref: String?,
        resourceHref: String?,
        resourceEtag: String?
    ) -> CalendarEvent? {
        guard let uid     = props["UID"],
              let rawSum  = props["SUMMARY"]
        else { return nil }

        let summary = unescape(rawSum)
        guard let start = parseDate(props, key: "DTSTART") else { return nil }

        let end: Date
        if let e = parseDate(props, key: "DTEND") {
            end = e
        } else if let durStr = props["DURATION"] {
            end = start.addingTimeInterval(parseDurationSeconds(durStr))
        } else {
            end = start.addingTimeInterval(3600)
        }

        return CalendarEvent(
            uid: uid,
            summary: summary,
            start: start,
            end: end,
            description: props["DESCRIPTION"].map { unescape($0) },
            location: props["LOCATION"].map { unescape($0) },
            calendarHref: calendarHref,
            etag: resourceEtag,
            href: resourceHref
        )
    }

    // ─── Task parser ──────────────────────────────────────────────────────────

    private static func parseTask(
        _ props: [String: String],
        calendarHref: String?,
        resourceHref: String?,
        resourceEtag: String?
    ) -> CalDavTask? {
        guard let uid     = props["UID"],
              let rawSum  = props["SUMMARY"]
        else { return nil }

        let summary = unescape(rawSum)

        var parentUid: String? = nil
        if let relatedTo = props["RELATED-TO"] {
            let rawLine = props["__RAW_RELATED-TO"] ?? ""
            if rawLine.uppercased().contains("RELTYPE=PARENT") {
                parentUid = relatedTo.split(separator: ",").first.map(String.init)
            }
        }

        return CalDavTask(
            uid: uid,
            summary: summary,
            description: props["DESCRIPTION"].map { unescape($0) },
            status: TaskStatus.fromIcal(props["STATUS"] ?? "NEEDS-ACTION"),
            priority: Int(props["PRIORITY"] ?? "") ?? 0,
            percentComplete: Int(props["PERCENT-COMPLETE"] ?? "") ?? 0,
            due: parseDate(props, key: "DUE"),
            dtstart: parseDate(props, key: "DTSTART"),
            completed: parseDate(props, key: "COMPLETED"),
            created: parseDate(props, key: "CREATED") ?? Date(),
            lastModified: parseDate(props, key: "LAST-MODIFIED") ?? Date(),
            categories: parseCategories(props["CATEGORIES"]),
            location: props["LOCATION"].map { unescape($0) },
            parentUid: parentUid,
            calendarHref: calendarHref,
            etag: resourceEtag,
            href: resourceHref
        )
    }

    // ─── Note (VJOURNAL) parser ───────────────────────────────────────────────

    private static func parseNote(
        _ props: [String: String],
        calendarHref: String?,
        resourceHref: String?,
        resourceEtag: String?
    ) -> Note? {
        guard let uid    = props["UID"],
              let rawSum = props["SUMMARY"]
        else { return nil }

        return Note(
            uid: uid,
            summary: unescape(rawSum),
            body: props["DESCRIPTION"].map { unescape($0) } ?? "",
            categories: parseCategories(props["CATEGORIES"]),
            created: parseDate(props, key: "DTSTAMP") ?? Date(),
            lastModified: parseDate(props, key: "LAST-MODIFIED") ?? Date(),
            calendarHref: calendarHref,
            etag: resourceEtag,
            href: resourceHref
        )
    }

    // ─── iCal serializers (domain → iCal text) ───────────────────────────────

    static func serializeTask(_ task: CalDavTask) -> String {
        var lines: [String] = []
        func line(_ s: String) { lines.append(s) }
        line("BEGIN:VCALENDAR")
        line("VERSION:2.0")
        line("PRODID:-//CrossDashboard//Native//EN")
        line("BEGIN:VTODO")
        line("UID:\(task.uid)")
        line("DTSTAMP:\(formatDate(Date()))")
        line("CREATED:\(formatDate(task.created))")
        line("LAST-MODIFIED:\(formatDate(Date()))")
        line("SUMMARY:\(escape(task.summary))")
        line("STATUS:\(task.status.rawValue)")
        line("PRIORITY:\(task.priority)")
        line("PERCENT-COMPLETE:\(task.percentComplete)")
        if let desc = task.description { line("DESCRIPTION:\(escape(desc))") }
        if let due  = task.due         { line("DUE:\(formatDate(due))") }
        if let dts  = task.dtstart     { line("DTSTART:\(formatDate(dts))") }
        if let comp = task.completed   { line("COMPLETED:\(formatDate(comp))") }
        if !task.categories.isEmpty    { line("CATEGORIES:\(task.categories.joined(separator: ","))") }
        if let loc  = task.location    { line("LOCATION:\(escape(loc))") }
        if let par  = task.parentUid   { line("RELATED-TO;RELTYPE=PARENT:\(par)") }
        line("END:VTODO")
        line("END:VCALENDAR")
        return lines.joined(separator: "\r\n") + "\r\n"
    }

    static func serializeNote(_ note: Note) -> String {
        var lines: [String] = []
        func line(_ s: String) { lines.append(s) }
        line("BEGIN:VCALENDAR")
        line("VERSION:2.0")
        line("PRODID:-//CrossDashboard//Native//EN")
        line("BEGIN:VJOURNAL")
        line("UID:\(note.uid)")
        line("DTSTAMP:\(formatDate(Date()))")
        line("LAST-MODIFIED:\(formatDate(Date()))")
        line("SUMMARY:\(escape(note.summary))")
        if !note.body.isEmpty           { line("DESCRIPTION:\(escape(note.body))") }
        if !note.categories.isEmpty     { line("CATEGORIES:\(note.categories.joined(separator: ","))") }
        line("END:VJOURNAL")
        line("END:VCALENDAR")
        return lines.joined(separator: "\r\n") + "\r\n"
    }

    static func serializeEvent(_ event: CalendarEvent) -> String {
        var lines: [String] = []
        func line(_ s: String) { lines.append(s) }
        line("BEGIN:VCALENDAR")
        line("VERSION:2.0")
        line("PRODID:-//CrossDashboard//Native//EN")
        line("BEGIN:VEVENT")
        line("UID:\(event.uid)")
        line("DTSTAMP:\(formatDate(Date()))")
        line("DTSTART:\(formatDate(event.start))")
        line("DTEND:\(formatDate(event.end))")
        line("SUMMARY:\(escape(event.summary))")
        if let desc = event.description { line("DESCRIPTION:\(escape(desc))") }
        if let loc  = event.location    { line("LOCATION:\(escape(loc))") }
        line("LAST-MODIFIED:\(formatDate(Date()))")
        line("END:VEVENT")
        line("END:VCALENDAR")
        return lines.joined(separator: "\r\n") + "\r\n"
    }

    // ─── Parsing helpers ─────────────────────────────────────────────────────

    private static func unfold(_ icalText: String) -> [String] {
        var result: [String] = []
        var current = ""
        let newline  = CharacterSet.newlines
        for line in icalText.components(separatedBy: .newlines) {
            if line.hasPrefix(" ") || line.hasPrefix("\t") {
                current += line.dropFirst()
            } else {
                if !current.isEmpty { result.append(current) }
                current = line
            }
        }
        if !current.isEmpty { result.append(current) }
        return result
    }

    // Formats: 20231015T143000Z or 20231015
    private static let utcFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        f.timeZone = TimeZone(identifier: "UTC")
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private static let localFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd'T'HHmmss"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private static let dateOnlyFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyyMMdd"
        f.timeZone = TimeZone(identifier: "UTC")
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private static func parseDate(_ props: [String: String], key: String) -> Date? {
        guard let rawLine = props["__RAW_\(key)"],
              let colonIdx = rawLine.firstIndex(of: ":") else { return nil }

        // Keep rawParamPart in original case so IANA timezone identifiers (e.g. "Europe/Berlin")
        // are preserved for TimeZone(identifier:), which is case-sensitive.
        let rawParamPart = String(rawLine[rawLine.startIndex ..< colonIdx])
        let paramPart    = rawParamPart.uppercased()   // uppercased only for keyword checks
        let value        = String(rawLine[rawLine.index(after: colonIdx)...]).trimmingCharacters(in: .whitespaces)

        // All-day: VALUE=DATE
        if paramPart.contains("VALUE=DATE") {
            dateOnlyFormatter.timeZone = .current
            return dateOnlyFormatter.date(from: String(value.prefix(8)))
        }

        // UTC Z suffix
        if value.hasSuffix("Z") {
            // Normalize to "yyyyMMdd'T'HHmmss'Z'" — handle compact form
            let compact = value.replacingOccurrences(of: "-", with: "")
                               .replacingOccurrences(of: ":", with: "")
            if compact.count >= 15 {
                let normalized = "\(compact.prefix(8))T\(compact.dropFirst(9).prefix(6))Z"
                return utcFormatter.date(from: normalized) ?? utcFormatter.date(from: value)
            }
            return utcFormatter.date(from: value)
        }

        // TZID param — search rawParamPart case-insensitively so the captured value retains
        // its original casing (e.g. "Europe/Berlin", not "EUROPE/BERLIN").
        // RFC 5545: a datetime with no Z and no TZID is "floating" and should be interpreted
        // in the local system timezone, not UTC.
        var tz: TimeZone = .current
        if paramPart.contains("TZID="),
           let tzRange = rawParamPart.range(of: "TZID=", options: .caseInsensitive) {
            let remaining = rawParamPart[tzRange.upperBound...]
            let tzId: String
            if let endIdx = remaining.firstIndex(where: { $0 == ";" || $0 == ":" }) {
                tzId = String(remaining[remaining.startIndex ..< endIdx])
            } else {
                tzId = String(remaining)
            }
            tz = TimeZone(identifier: tzId) ?? .current
        }

        localFormatter.timeZone = tz
        return localFormatter.date(from: value)
    }

    private static func parseDurationSeconds(_ s: String) -> TimeInterval {
        var total: TimeInterval = 0
        let pattern = /P(?:(\d+)W)?(?:(\d+)D)?(?:T(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?)?/
        if let match = s.firstMatch(of: pattern) {
            let weeks   = Double(match.output.1 ?? Substring("")) ?? 0
            let days    = Double(match.output.2 ?? Substring("")) ?? 0
            let hours   = Double(match.output.3 ?? Substring("")) ?? 0
            let minutes = Double(match.output.4 ?? Substring("")) ?? 0
            let seconds = Double(match.output.5 ?? Substring("")) ?? 0
            total = weeks * 7 * 86400 + days * 86400 + hours * 3600 + minutes * 60 + seconds
        }
        return total > 0 ? total : 3600
    }

    private static func parseCategories(_ raw: String?) -> [String] {
        guard let raw, !raw.isEmpty else { return [] }
        return raw.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
    }

    static func unescape(_ value: String) -> String {
        value
            .replacingOccurrences(of: "\\n", with: "\n")
            .replacingOccurrences(of: "\\N", with: "\n")
            .replacingOccurrences(of: "\\,", with: ",")
            .replacingOccurrences(of: "\\;", with: ";")
            .replacingOccurrences(of: "\\\\", with: "\\")
    }

    private static func escape(_ value: String) -> String {
        value
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: ",",  with: "\\,")
            .replacingOccurrences(of: ";",  with: "\\;")
    }

    static func formatDate(_ date: Date) -> String {
        utcFormatter.string(from: date)
    }
}
